src/main/java/io/odpf/firehose/sink/mongodb/MongoSink.java

```diff
@@ -0,0 +1,125 @@
+        this.mongoCollection = mongoCollection;
+        this.mongoRequestHandler = mongoRequestHandler;
+        this.mongoClient = mongoClient;
+        this.mongoRetryStatusCodeBlacklist = mongoRetryStatusCodeBlacklist;
+    }
+
+    @Override
+    protected void prepare(List<Message> messageList) {
+        this.messages = messageList;
+        bulkRequest.clear();
+
+        messages.forEach(message -> bulkRequest.add(mongoRequestHandler.getRequest(message)));
+    }
+
+    @Override
+    protected List<Message> execute() throws Exception {
+        List<BulkWriteError> bulkWriteErrors = processRequest();
+        if (!bulkWriteErrors.isEmpty()) {
+            getInstrumentation().logWarn("Bulk request failed");
+
+            handleBulkWriteErrors(bulkWriteErrors);
+        }
+
+        List<Message> failedMessages = new ArrayList<>();
+
+        bulkWriteErrors.forEach((bulkWriteError) -> {
+            if (!mongoRetryStatusCodeBlacklist.contains(String.valueOf(bulkWriteError.getCode()))) {
+                failedMessages.add(messages.get(bulkWriteError.getIndex()));
+            }
+        });
+        return failedMessages;
+    }
+
+    @Override
+    public void close() throws IOException {
+        getInstrumentation().logInfo("MongoDB connection closing");
+        this.mongoClient.close();
+    }
+
+    /**
+     * Process request list.
+     *
+     * @return the list
+     */
+    List<BulkWriteError> processRequest() {
+        List<BulkWriteError> bulkWriteErrors = new ArrayList<>();
+        try {
+            handleBulkWriteResult(mongoCollection.bulkWrite(bulkRequest));
+        } catch (BulkWriteException bulkWriteException) {
+
+            bulkWriteErrors = bulkWriteException.getWriteErrors();
+        }
+
+        return bulkWriteErrors;
+    }
+
+    private void handleBulkWriteResult(BulkWriteResult bulkWriteResult) {
+
+        getInstrumentation().logInfo("Successfully inserted {} documents", bulkWriteResult.getInsertedCount());
+        getInstrumentation().logInfo("Successfully updated {} documents", bulkWriteResult.getModifiedCount());
+
+    }
+
+    private void handleBulkWriteErrors(List<BulkWriteError> bulkWriteErrors) throws NeedToRetry {
+        int failedResponseCount = 0;
+        for (BulkWriteError bulkWriteError : bulkWriteErrors) {
+            failedResponseCount++;
+            String responseStatus = String.valueOf(bulkWriteError.getCode());
+            if (mongoRetryStatusCodeBlacklist.contains(responseStatus)) {
+                getInstrumentation().logInfo("Not retrying due to response status: {} is under blacklisted status code", responseStatus);
+                getInstrumentation().incrementCounterWithTags(SINK_MESSAGES_DROP_TOTAL, "cause=" + bulkWriteError.getMessage());
+                getInstrumentation().logInfo("Message dropped because of status code: " + responseStatus);
+            } else {
+                throw new NeedToRetry(String.valueOf(bulkWriteError.getCode()));
```

Comment:
let's not throw this.. we should be returning messages which are failed. Decorator will take care of retries. We will not use this in future.

Reply:
fixed